biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Indexer.java

```diff
@@ -109,21 +110,38 @@ File get(String bsn, Version version, Map<String, String> properties, DownloadLi
 			return null;
 
 		URI url = contentCapability.url();
+		boolean packed = url.getPath()
+			.endsWith(Unpack200.PACKED_SUFFIX);
 
-		final File source = client.getCacheFileFor(url);
 		final File link = new File(location, bsn + "-" + version + ".jar");
 
-		IO.createSymbolicLinkOrCopy(link, source);
-
 		Promise<File> go = client.build()
 			.useCache(MAX_STALE)
-			.async(url.toURL())
-			.map(file -> link);
+			.async(url.toURL());
 
-		if (listeners.length == 0)
-			return go.getValue();
+		if (listeners.length == 0) {
+			return handleDownloadResult(packed, link, go);
+		}
 
 		new DownloadListenerPromise(reporter, name + ": get " + bsn + ";" + version + " " + url, go, listeners);
+
+		return handleDownloadResult(packed, link, go);
```

Comment:
You can't do it this way. You need to do any unpacking as a map operation on the promise. Otherwise you block here in this method and the point of using listeners is to NOT block here.